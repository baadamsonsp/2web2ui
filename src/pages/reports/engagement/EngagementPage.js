import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import { Page } from '@sparkpost/matchbox';
import { refreshEngagementReport } from 'src/actions/engagementReport';
import { selectReportSearchOptions } from 'src/selectors/reportSearchOptions';
import ReportOptions from 'src/pages/reports/components/ReportOptions';
import EngagementChart from './components/EngagementChart';
import EngagementSummary from './components/EngagementSummary';
import EngagementTable from './components/EngagementTable';

export class EngagementPage extends Component {

  componentDidUpdate(prevProps) {
    if (!_.isEqual(prevProps.reportOptions, this.props.reportOptions)) {
      this.props.refreshEngagementReport(this.props.reportOptions);
    }
  }

  render() {
    const { loading, aggregateMetrics, linkMetrics, engagementSearchOptions } = this.props;

    return (
      <Page title='Engagement Report'>
        <ReportOptions reportLoading={loading} searchOptions={engagementSearchOptions} />
        <EngagementSummary
          accepted={aggregateMetrics.data.count_accepted}
          clicks={aggregateMetrics.data.count_unique_clicked_approx}
          loading={aggregateMetrics.loading}
          opens={aggregateMetrics.data.count_unique_confirmed_opened_approx}
          sent={aggregateMetrics.data.count_sent}
        />
        <EngagementChart
          accepted={aggregateMetrics.data.count_accepted}
          clicks={aggregateMetrics.data.count_unique_clicked_approx}
          loading={aggregateMetrics.loading}
          opens={aggregateMetrics.data.count_unique_confirmed_opened_approx}
          sent={aggregateMetrics.data.count_sent}
        />
        <EngagementTable data={linkMetrics.data} loading={linkMetrics.loading} />
      </Page>
    );
  }
}

const mapStateToProps = (state) => ({
  loading: state.engagementReport.loading,
  aggregateMetrics: state.engagementReport.aggregateMetrics,
  linkMetrics: state.engagementReport.linkMetrics,
  reportOptions: state.reportOptions,
  engagementSearchOptions: selectReportSearchOptions(state)
});
const mapDispatchToProps = { refreshEngagementReport };

export default connect(mapStateToProps, mapDispatchToProps)(EngagementPage);
