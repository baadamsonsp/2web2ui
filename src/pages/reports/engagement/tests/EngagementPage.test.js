import React from 'react';
import { shallow } from 'enzyme';
import { EngagementPage } from '../EngagementPage';

describe('Engagement Report Page', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      loading: false,
      aggregateMetrics: {
        data: {
          count_accepted: 1,
          count_sent: 2,
          count_unique_clicked_approx: 3,
          count_unique_confirmed_opened_approx: 4
        },
        loading: false
      },
      linkMetrics: {
        data: [
          { count_clicked: 1776, count_raw_clicked_approx: 1657, link_name: 'Raw URL' }
        ],
        loading: false
      },
      refreshEngagementReport: jest.fn(),
      reportOptions: {},
      engagementSearchOptions: {}
    };
    wrapper = shallow(<EngagementPage {...props} />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should refresh when report options reference changes', () => {
    wrapper.setProps({ reportOptions: { relativeRange: 'day' }});
    expect(props.refreshEngagementReport).toHaveBeenLastCalledWith({ relativeRange: 'day' });
  });

  it('should not refresh when report options reference is unchanged', () => {
    wrapper.setProps({ reportOptions: {}});
    expect(props.refreshEngagementReport).toHaveBeenCalledTimes(0);
  });

});
