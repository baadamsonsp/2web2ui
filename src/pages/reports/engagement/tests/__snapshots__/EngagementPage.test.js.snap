// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engagement Report Page should render 1`] = `
<Page
  empty={Object {}}
  title="Engagement Report"
>
  <withRouter(Connect(ReportOptions))
    reportLoading={false}
    searchOptions={Object {}}
  />
  <Connect(EngagementSummary)
    accepted={1}
    clicks={3}
    loading={false}
    opens={4}
    sent={2}
  />
  <LoadableComponent
    accepted={1}
    clicks={3}
    loading={false}
    opens={4}
    sent={2}
  />
  <EngagementTable
    data={
      Array [
        Object {
          "count_clicked": 1776,
          "count_raw_clicked_approx": 1657,
          "link_name": "Raw URL",
        },
      ]
    }
    loading={false}
  />
</Page>
`;
